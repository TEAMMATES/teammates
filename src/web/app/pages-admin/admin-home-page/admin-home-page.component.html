<div class="card bg-light top-padded">
  <div class="card-body">
    <div>
      <strong>Adding Multiple Instructors</strong>
      <p class="text-muted">
        <i class="fas fa-exclamation-circle"></i> Add Instructor Details in the format: Name | Email | Institution
      </p>
      <textarea class="form-control" id="instructor-details-single-line" [(ngModel)]="instructorDetails" aria-label="Add multiple instructors"></textarea>
    </div>

    <div class="top-padded">
      <button class="btn btn-primary" id="add-instructor-single-line" (click)="validateAndAddInstructorDetails()">Add Instructors</button>
    </div>
  </div>
</div>

<div class="card bg-light top-padded">
  <div class="card-body">
    <div>
      <strong>Adding a Single Instructor</strong>
    </div>

    <div class="top-padded">
      <label for="instructor-name">
        <strong>Name:</strong>
      </label>
      <input class="form-control" type="text" id="instructor-name" [(ngModel)]="instructorName">
    </div>

    <div class="top-padded">
      <label for="instructor-email">
        <strong>Email:</strong>
      </label>
      <input class="form-control" type="text" id="instructor-email" [(ngModel)]="instructorEmail">
    </div>

    <div class="top-padded">
      <label for="instructor-institution">
        <strong>Institution:</strong>
      </label>
      <input class="form-control" type="text" id="instructor-institution" [(ngModel)]="instructorInstitution">
    </div>

    <div class="top-padded">
      <button class="btn btn-primary" id="add-instructor" (click)="validateAndAddInstructorDetail()">Add Instructor</button>
    </div>
  </div>
</div>

<tm-account-request-table *ngIf="accountReqs.length" [accountRequests]="accountReqs" [searchString]=""></tm-account-request-table>
