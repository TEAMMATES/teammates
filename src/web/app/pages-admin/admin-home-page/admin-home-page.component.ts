import { Component, OnInit } from '@angular/core';
import { Observable, of } from 'rxjs';
import { AccountService } from '../../../services/account.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { TimezoneService } from '../../../services/timezone.service';
import { AccountRequests } from '../../../types/api-output';
import { AccountCreateRequest } from '../../../types/api-request';
import { AccountRequestTableRowModel } from '../../components/account-requests-table/account-request-table-model';
import { FormatDateDetailPipe } from '../../components/teammates-common/format-date-detail.pipe';
import { ErrorMessageOutput } from '../../error-message-output';

/**
 * Admin home page.
 */
@Component({
  selector: 'tm-admin-home-page',
  templateUrl: './admin-home-page.component.html',
  styleUrls: ['./admin-home-page.component.scss'],
})
export class AdminHomePageComponent implements OnInit {

  instructorDetails: string = '';
  instructorName: string = '';
  instructorEmail: string = '';
  instructorInstitution: string = '';

  accountReqs: AccountRequestTableRowModel[] = [];
  activeRequests: number = 0;
  currentPage: number = 1;
  pageSize: number = 20;
  items$: Observable<any> = of([]);

  constructor(
    private accountService: AccountService,
    private statusMessageService: StatusMessageService,
    private timezoneService: TimezoneService,
    private formatDateDetailPipe: FormatDateDetailPipe,
  ) {}

  ngOnInit(): void {
    this.fetchAccountRequests();
  }

  /**
   * Validates and adds the instructor details filled with first form.
   */
  validateAndAddInstructorDetails(): Promise<void> {
    const lines: string[] = this.instructorDetails.split(/\r?\n/);
    const invalidLines: string[] = [];
    const accountRequests: Promise<void>[] = [];
    for (const line of lines) {
      const instructorDetailsSplit: string[] = line.split(/[|\t]/).map((item: string) => item.trim());
      let errorMessage = '';
      switch (instructorDetailsSplit.length) {
        case 1:
          // Triggers when instructorDetails is empty
          if (!instructorDetailsSplit[0]) {
            continue;
          }

          invalidLines.push(line);
          this.statusMessageService.showErrorToast('email cannot be null');
          continue;
        case 2:
          invalidLines.push(line);
          this.statusMessageService.showErrorToast('institution cannot be null');
          continue;
        case 3:
          break;
        default:
          invalidLines.push(line);
          this.statusMessageService.showErrorToast('Too many fields, please check that all lines contains only a '
            + 'name, email and institution.');
          continue;
      }

      // Update the character numbers if it changes in the backend.
      if (!instructorDetailsSplit[0]) {
        errorMessage = 'The field \'person name\' is empty. '
        + 'The value of a/an person name should be no longer than 100 characters. '
        + 'It should not be empty.\n';
      }
      if (!instructorDetailsSplit[1]) {
        errorMessage += 'The field \'email\' is empty. '
        + 'An email address contains some text followed by one \'@\' sign followed by some more text, '
        + 'and should end with a top level domain address like .com. '
        + 'It cannot be longer than 254 characters, cannot be empty and cannot contain spaces.\n';
      }
      if (!instructorDetailsSplit[2]) {
        errorMessage += 'The field \'institute name\' is empty. '
        + 'The value of a/an institute name should be no longer than 128 characters. '
        + 'It should not be empty.';
      }
      if (errorMessage !== '') {
        invalidLines.push(line);
        this.statusMessageService.showErrorToast(errorMessage);
        continue;
      }

      const requestData: AccountCreateRequest = {
        instructorEmail: instructorDetailsSplit[1],
        instructorName: instructorDetailsSplit[0],
        instructorInstitution: instructorDetailsSplit[2],
      };

      const newRequest: Promise<void> = new Promise((resolve, reject) => {
        this.accountService.createAccountRequest(requestData)
        .subscribe({
          next: () => {
            resolve();
          },
          error: (resp: ErrorMessageOutput) => {
            invalidLines.push(line);
            this.statusMessageService.showErrorToast(resp.error.message);
            reject();
          },
        });
      });

      accountRequests.push(newRequest);
    }

    return Promise.allSettled(accountRequests).then(() => {
      this.instructorDetails = invalidLines.join('\r\n');
      this.fetchAccountRequests();
    });
  }

  /**
   * Validates and adds the instructor detail filled with second form.
   */
  validateAndAddInstructorDetail(): void {
    if (!this.instructorName || !this.instructorEmail || !this.instructorInstitution) {
      // TODO handle error
      return;
    }

    const requestData: AccountCreateRequest = {
      instructorEmail: this.instructorEmail,
      instructorName: this.instructorName,
      instructorInstitution: this.instructorInstitution,
    };

    this.accountService.createAccountRequest(requestData)
    .subscribe({
      next: () => {
        this.instructorName = '';
        this.instructorEmail = '';
        this.instructorInstitution = '';

        this.fetchAccountRequests();
      },
      error: (resp: ErrorMessageOutput) => {
        this.statusMessageService.showErrorToast(resp.error.message);
      },
    });
  }

  private formatAccountRequests(requests: AccountRequests): AccountRequestTableRowModel[] {
    const timezone: string = this.timezoneService.guessTimezone() || 'UTC';
    return requests.accountRequests.map((request) => {
      return {
        id: request.id,
        name: request.name,
        email: request.email,
        status: request.status,
        instituteAndCountry: request.institute,
        createdAtText: this.formatDateDetailPipe.transform(request.createdAt, timezone),
        registeredAtText: request.registeredAt
        ? this.formatDateDetailPipe.transform(request.registeredAt, timezone) : '',
        comments: request.comments || '',
        registrationLink: '',
        showLinks: false,
      };
    });
  }

  fetchAccountRequests(): void {
    this.accountService.getPendingAccountRequests().subscribe({
      next: (resp: AccountRequests) => {
        this.accountReqs = this.formatAccountRequests(resp);
      },
      error: (resp: ErrorMessageOutput) => {
        this.statusMessageService.showErrorToast(resp.error.message);
      },
    });
  }
}
