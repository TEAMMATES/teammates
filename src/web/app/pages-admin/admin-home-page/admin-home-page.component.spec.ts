import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { FormsModule } from '@angular/forms';
import { RouterTestingModule } from '@angular/router/testing';
import { of, throwError } from 'rxjs';
import { AdminHomePageComponent } from './admin-home-page.component';
import { NewInstructorDataRowComponent } from './new-instructor-data-row/new-instructor-data-row.component';
import { AccountService } from '../../../services/account.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { createBuilder } from '../../../test-helpers/generic-builder';
import { AccountRequest, AccountRequests, AccountRequestStatus, MessageOutput } from '../../../types/api-output';
import { AccountCreateRequest } from '../../../types/api-request';
import { AccountRequestTableRowModel } from '../../components/account-requests-table/account-request-table-model';
import { AccountRequestTableModule } from '../../components/account-requests-table/account-request-table.module';
import { AjaxLoadingModule } from '../../components/ajax-loading/ajax-loading.module';
import { LoadingSpinnerModule } from '../../components/loading-spinner/loading-spinner.module';
import { FormatDateDetailPipe } from '../../components/teammates-common/format-date-detail.pipe';
import { ErrorMessageOutput } from '../../error-message-output';

const accountCreateRequestBuilder = createBuilder<AccountCreateRequest>({
  instructorEmail: '',
  instructorName: '',
  instructorInstitution: '',
});

const accountRequestBuilder = createBuilder<AccountRequest>({
  id: '',
  email: '',
  name: '',
  institute: '',
  registrationKey: '',
  status: AccountRequestStatus.PENDING,
  createdAt: 0,
});

const messageOutputBuilder = createBuilder<MessageOutput>({
  message: '',
});

const errorMessageOutputBuilder = createBuilder<ErrorMessageOutput>({
  error: messageOutputBuilder.build(),
  status: 0,
});

const accountRequestTableRowModelBuilder = createBuilder<AccountRequestTableRowModel>({
  id: '',
  name: '',
  email: '',
  status: AccountRequestStatus.PENDING,
  instituteAndCountry: '',
  createdAtText: '',
  registeredAtText: '',
  comments: '',
  registrationLink: '',
  showLinks: false,
});

describe('AdminHomePageComponent', () => {
  let component: AdminHomePageComponent;
  let fixture: ComponentFixture<AdminHomePageComponent>;
  let accountService: AccountService;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [
        AdminHomePageComponent,
        NewInstructorDataRowComponent,
      ],
      imports: [
        FormsModule,
        HttpClientTestingModule,
        LoadingSpinnerModule,
        AccountRequestTableModule,
        AjaxLoadingModule,
        RouterTestingModule,
      ],
      providers: [
        AccountService,
        FormatDateDetailPipe,
        StatusMessageService,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(AdminHomePageComponent);
    accountService = TestBed.inject(AccountService);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('validateAndAddInstructorDetail: should create one instructor account request if all fields are filled', () => {
    component.instructorName = 'Instructor Name';
    component.instructorEmail = 'instructor@example.com';
    component.instructorInstitution = 'Instructor Institution';
    fixture.detectChanges();

    const accountCreateRequest = accountCreateRequestBuilder
      .instructorName('Instructor Name')
      .instructorEmail('instructor@example.com')
      .instructorInstitution('Instructor Institution')
      .build();

    const createAccountRequestSpy = jest.spyOn(accountService, 'createAccountRequest')
      .mockReturnValue(of(accountRequestBuilder.build()));

    const button: any = fixture.debugElement.nativeElement.querySelector('#add-instructor');
    button.click();

    expect(createAccountRequestSpy).toHaveBeenCalledTimes(1);
    expect(createAccountRequestSpy).toHaveBeenCalledWith(accountCreateRequest);

    // Clear instructor fields
    expect(component.instructorName).toEqual('');
    expect(component.instructorEmail).toEqual('');
    expect(component.instructorInstitution).toEqual('');
  });

  it('validateAndAddInstructorDetail: should not create one instructor account request '
  + 'if some fields are empty', () => {
    const button: any = fixture.debugElement.nativeElement.querySelector('#add-instructor');
    const createAccountRequestSpy = jest.spyOn(accountService, 'createAccountRequest');

    component.instructorName = '';
    component.instructorEmail = 'instructor@example.com';
    component.instructorInstitution = 'Instructor Institution';
    fixture.detectChanges();

    button.click();

    expect(component.instructorName).toEqual('');
    expect(component.instructorEmail).toEqual('instructor@example.com');
    expect(component.instructorInstitution).toEqual('Instructor Institution');
    expect(createAccountRequestSpy).not.toHaveBeenCalled();

    component.instructorName = 'Instructor Name';
    component.instructorEmail = '';
    component.instructorInstitution = 'Instructor Institution';
    fixture.detectChanges();

    button.click();

    expect(component.instructorName).toEqual('Instructor Name');
    expect(component.instructorEmail).toEqual('');
    expect(component.instructorInstitution).toEqual('Instructor Institution');
    expect(createAccountRequestSpy).not.toHaveBeenCalled();

    component.instructorName = 'Instructor Name';
    component.instructorEmail = 'instructor@example.com';
    component.instructorInstitution = '';
    fixture.detectChanges();

    button.click();

    expect(component.instructorName).toEqual('Instructor Name');
    expect(component.instructorEmail).toEqual('instructor@example.com');
    expect(component.instructorInstitution).toEqual('');
    expect(createAccountRequestSpy).not.toHaveBeenCalled();
  });

  it('validateAndAddInstructorDetails: should only create account requests for valid instructor details '
  + 'when there are invalid lines in the single line field', () => {
    component.instructorDetails = [
      'Instructor A | instructora@example.com | Institution A',
      'Instructor B | instructorb@example.com',
      'Instructor C | | instructorc@example.com',
      'Instructor D | instructord@example.com | Institution D',
      '| instructore@example.com | Institution E',
    ].join('\n');
    fixture.detectChanges();

    const accountCreateRequestA = accountCreateRequestBuilder
      .instructorName('Instructor A')
      .instructorEmail('instructora@example.com')
      .instructorInstitution('Institution A')
      .build();
    const accountCreateRequestD = accountCreateRequestBuilder
      .instructorName('Instructor D')
      .instructorEmail('instructord@example.com')
      .instructorInstitution('Institution D')
      .build();

    const createAccountRequestSpy = jest.spyOn(accountService, 'createAccountRequest')
      .mockImplementation((request) => {
        switch (request.instructorEmail) {
          case accountCreateRequestA.instructorEmail:
            return of(accountRequestBuilder.build());
          case accountCreateRequestD.instructorEmail:
            return of(accountRequestBuilder.build());
          default:
            return throwError(() => errorMessageOutputBuilder.build());
        }
      });

    const button: any = fixture.debugElement.nativeElement.querySelector('#add-instructor-single-line');
    button.click();

    expect(component.instructorDetails).toEqual([
      'Instructor B | instructorb@example.com',
      'Instructor C | | instructorc@example.com',
      '| instructore@example.com | Institution E',
    ].join('\r\n'));

    expect(createAccountRequestSpy).toHaveBeenCalledTimes(2);
    expect(createAccountRequestSpy).toHaveBeenNthCalledWith(1, accountCreateRequestA);
    expect(createAccountRequestSpy).toHaveBeenNthCalledWith(2, accountCreateRequestD);
  });

  it('should snap with default view', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with some instructors details', () => {
    component.accountReqs = [
      accountRequestTableRowModelBuilder
        .name('Instructor A')
        .email('instructora@example.com')
        .status(AccountRequestStatus.PENDING)
        .instituteAndCountry('Institution and Country A')
        .createdAtText('Created Time A')
        .comments('Comment A')
        .build(),

      accountRequestTableRowModelBuilder
        .name('Instructor B')
        .email('instructorb@example.com')
        .status(AccountRequestStatus.APPROVED)
        .instituteAndCountry('Institution and Country B')
        .createdAtText('Created Time B')
        .comments('Comment B')
        .build(),

      accountRequestTableRowModelBuilder
        .name('Instructor C')
        .email('instructorc@example.com')
        .status(AccountRequestStatus.REJECTED)
        .instituteAndCountry('Institution and Country C')
        .createdAtText('Created Time C')
        .comments('Comment C')
        .build(),
    ];
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('validateAndAddInstructorDetails: should create multiple instructor account requests when split by tabs', () => {
    component.instructorDetails =
      `Instructor A   \t  instructora@example.com \t  Institution A\n
      Instructor B \t instructorb@example.com \t Institution B`;
    fixture.detectChanges();

    const accountCreateRequestA = accountCreateRequestBuilder
      .instructorName('Instructor A')
      .instructorEmail('instructora@example.com')
      .instructorInstitution('Institution A')
      .build();
    const accountCreateRequestB = accountCreateRequestBuilder
      .instructorName('Instructor B')
      .instructorEmail('instructorb@example.com')
      .instructorInstitution('Institution B')
      .build();

    const createAccountRequestSpy = jest.spyOn(accountService, 'createAccountRequest')
      .mockImplementation((request) => {
        switch (request.instructorEmail) {
          case accountCreateRequestA.instructorEmail:
            return of(accountRequestBuilder.build());
          case accountCreateRequestB.instructorEmail:
            return of(accountRequestBuilder.build());
          default:
            return throwError(() => errorMessageOutputBuilder.build());
        }
      });

    const button: any = fixture.debugElement.nativeElement.querySelector('#add-instructor-single-line');
    button.click();

    expect(component.instructorDetails).toEqual('');

    expect(createAccountRequestSpy).toHaveBeenNthCalledWith(1, accountCreateRequestA);
    expect(createAccountRequestSpy).toHaveBeenNthCalledWith(2, accountCreateRequestB);
  });

  it('validateAndAddInstructorDetails: should create multiple instructor account requests '
  + 'when split by vertical bars', () => {
    component.instructorDetails =
      `Instructor A | instructora@example.com | Institution A\n
      Instructor B | instructorb@example.com | Institution B`;
    fixture.detectChanges();

    const accountCreateRequestA = accountCreateRequestBuilder
      .instructorName('Instructor A')
      .instructorEmail('instructora@example.com')
      .instructorInstitution('Institution A')
      .build();
    const accountCreateRequestB = accountCreateRequestBuilder
      .instructorName('Instructor B')
      .instructorEmail('instructorb@example.com')
      .instructorInstitution('Institution B')
      .build();

    const createAccountRequestSpy = jest.spyOn(accountService, 'createAccountRequest')
      .mockImplementation((request) => {
        switch (request.instructorEmail) {
          case accountCreateRequestA.instructorEmail:
            return of(accountRequestBuilder.build());
          case accountCreateRequestB.instructorEmail:
            return of(accountRequestBuilder.build());
          default:
            return throwError(() => errorMessageOutputBuilder.build());
        }
      });

    const button: any = fixture.debugElement.nativeElement.querySelector('#add-instructor-single-line');
    button.click();

    expect(component.instructorDetails).toEqual('');

    expect(createAccountRequestSpy).toHaveBeenNthCalledWith(1, accountCreateRequestA);
    expect(createAccountRequestSpy).toHaveBeenNthCalledWith(2, accountCreateRequestB);
  });

  it('fetchAccountRequests: should update account requests binding if pending account requests exist', () => {
    const accountRequestA = accountRequestBuilder
      .name('Instructor A')
      .email('instructora@example.com')
      .institute('Institution A')
      .build();
    const accountRequestB = accountRequestBuilder
      .name('Instructor B')
      .email('instructorb@example.com')
      .institute('Institution B')
      .build();
    const accountRequests: AccountRequests = {
      accountRequests: [
        accountRequestA,
        accountRequestB,
      ],
    };

    jest.spyOn(accountService, 'getPendingAccountRequests')
      .mockReturnValue(of(accountRequests));

    component.fetchAccountRequests();

    expect(component.accountReqs.length).toEqual(2);

    expect(component.accountReqs[0].name).toEqual('Instructor A');
    expect(component.accountReqs[0].email).toEqual('instructora@example.com');
    expect(component.accountReqs[0].instituteAndCountry).toEqual('Institution A');

    expect(component.accountReqs[1].name).toEqual('Instructor B');
    expect(component.accountReqs[1].email).toEqual('instructorb@example.com');
    expect(component.accountReqs[1].instituteAndCountry).toEqual('Institution B');
  });

  it('fetchAccountRequests: should not update account requests binding if no pending account requests exist', () => {
    const accountRequests: AccountRequests = {
      accountRequests: [],
    };

    jest.spyOn(accountService, 'getPendingAccountRequests')
      .mockReturnValue(of(accountRequests));

    component.fetchAccountRequests();

    expect(component.accountReqs.length).toEqual(0);
  });
});
