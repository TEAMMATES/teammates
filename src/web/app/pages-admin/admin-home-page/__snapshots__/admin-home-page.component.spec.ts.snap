// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminHomePageComponent should snap with default view 1`] = `
<tm-admin-home-page
  accountReqs={[Function Array]}
  accountService={[Function AccountService]}
  activeRequests="0"
  currentPage={[Function Number]}
  formatDateDetailPipe={[Function FormatDateDetailPipe]}
  instructorDetails=""
  instructorEmail=""
  instructorInstitution=""
  instructorName=""
  items$={[Function Observable]}
  pageSize={[Function Number]}
  statusMessageService={[Function StatusMessageService]}
  timezoneService={[Function TimezoneService]}
>
  <div
    class="card bg-light top-padded"
  >
    <div
      class="card-body"
    >
      <div>
        <strong>
          Adding Multiple Instructors
        </strong>
        <p
          class="text-muted"
        >
          <i
            class="fas fa-exclamation-circle"
          />
           Add Instructor Details in the format: Name | Email | Institution 
        </p>
        <textarea
          aria-label="Add multiple instructors"
          class="form-control ng-untouched ng-pristine ng-valid"
          id="instructor-details-single-line"
        />
      </div>
      <div
        class="top-padded"
      >
        <button
          class="btn btn-primary"
          id="add-instructor-single-line"
        >
          Add Instructors
        </button>
      </div>
    </div>
  </div><div
    class="card bg-light top-padded"
  >
    <div
      class="card-body"
    >
      <div>
        <strong>
          Adding a Single Instructor
        </strong>
      </div>
      <div
        class="top-padded"
      >
        <label
          for="instructor-name"
        >
          <strong>
            Name:
          </strong>
        </label>
        <input
          class="form-control ng-untouched ng-pristine ng-valid"
          id="instructor-name"
          type="text"
        />
      </div>
      <div
        class="top-padded"
      >
        <label
          for="instructor-email"
        >
          <strong>
            Email:
          </strong>
        </label>
        <input
          class="form-control ng-untouched ng-pristine ng-valid"
          id="instructor-email"
          type="text"
        />
      </div>
      <div
        class="top-padded"
      >
        <label
          for="instructor-institution"
        >
          <strong>
            Institution:
          </strong>
        </label>
        <input
          class="form-control ng-untouched ng-pristine ng-valid"
          id="instructor-institution"
          type="text"
        />
      </div>
      <div
        class="top-padded"
      >
        <button
          class="btn btn-primary"
          id="add-instructor"
        >
          Add Instructor
        </button>
      </div>
    </div>
  </div>
</tm-admin-home-page>
`;

exports[`AdminHomePageComponent should snap with some instructors details 1`] = `
<tm-admin-home-page
  accountReqs={[Function Array]}
  accountService={[Function AccountService]}
  activeRequests="0"
  currentPage={[Function Number]}
  formatDateDetailPipe={[Function FormatDateDetailPipe]}
  instructorDetails=""
  instructorEmail=""
  instructorInstitution=""
  instructorName=""
  items$={[Function Observable]}
  pageSize={[Function Number]}
  statusMessageService={[Function StatusMessageService]}
  timezoneService={[Function TimezoneService]}
>
  <div
    class="card bg-light top-padded"
  >
    <div
      class="card-body"
    >
      <div>
        <strong>
          Adding Multiple Instructors
        </strong>
        <p
          class="text-muted"
        >
          <i
            class="fas fa-exclamation-circle"
          />
           Add Instructor Details in the format: Name | Email | Institution 
        </p>
        <textarea
          aria-label="Add multiple instructors"
          class="form-control ng-untouched ng-pristine ng-valid"
          id="instructor-details-single-line"
        />
      </div>
      <div
        class="top-padded"
      >
        <button
          class="btn btn-primary"
          id="add-instructor-single-line"
        >
          Add Instructors
        </button>
      </div>
    </div>
  </div><div
    class="card bg-light top-padded"
  >
    <div
      class="card-body"
    >
      <div>
        <strong>
          Adding a Single Instructor
        </strong>
      </div>
      <div
        class="top-padded"
      >
        <label
          for="instructor-name"
        >
          <strong>
            Name:
          </strong>
        </label>
        <input
          class="form-control ng-untouched ng-pristine ng-valid"
          id="instructor-name"
          type="text"
        />
      </div>
      <div
        class="top-padded"
      >
        <label
          for="instructor-email"
        >
          <strong>
            Email:
          </strong>
        </label>
        <input
          class="form-control ng-untouched ng-pristine ng-valid"
          id="instructor-email"
          type="text"
        />
      </div>
      <div
        class="top-padded"
      >
        <label
          for="instructor-institution"
        >
          <strong>
            Institution:
          </strong>
        </label>
        <input
          class="form-control ng-untouched ng-pristine ng-valid"
          id="instructor-institution"
          type="text"
        />
      </div>
      <div
        class="top-padded"
      >
        <button
          class="btn btn-primary"
          id="add-instructor"
        >
          Add Instructor
        </button>
      </div>
    </div>
  </div><tm-account-request-table>
    <div
      class="card bg-light top-padded"
    >
      <div
        class="card-header bg-primary text-white"
      >
        <strong>
          Pending Account Requests
        </strong>
      </div>
      <div
        class="table-responsive"
      >
        <table
          class="table table-striped data-table"
          id="search-table-account-request"
        >
          <thead>
            <tr>
              <th>
                Name
              </th>
              <th>
                Email
              </th>
              <th>
                Status
              </th>
              <th>
                Institute, Country
              </th>
              <th>
                Created At
              </th>
              <th>
                Comments
              </th>
              <th>
                Options
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                Instructor A
              </td>
              <td>
                instructora@example.com
              </td>
              <td>
                PENDING
              </td>
              <td>
                Institution and Country A
              </td>
              <td
                id="timestamp-box"
              >
                Created Time A
              </td>
              <td>
                <div
                  id="comment-box"
                >
                  Comment A
                </div>
              </td>
              <td
                class="align-middle"
              >
                <div
                  class="d-flex flex-row align-items-center justify-content-center gap-2"
                >
                  <div>
                    <a
                      id="edit-account-request-0"
                    >
                      <i
                        class="fa-solid fa-pen"
                      />
                    </a>
                  </div>
                  <div
                    class="ngb-tooltip-class"
                    placement="top"
                  >
                    <a
                      id="delete-account-request-0"
                    >
                      <i
                        class="fa-solid fa-trash"
                      />
                    </a>
                  </div>
                  <div>
                    <a
                      id="view-account-request-0"
                    >
                      <i
                        class="fa-solid fa-eye"
                      />
                    </a>
                  </div>
                  <button
                    class="btn btn-success"
                    id="approve-account-request-0"
                  >
                    Approve
                  </button>
                  <span
                    class="dropdown"
                    container="body"
                    ngbdropdown=""
                  >
                    <button
                      aria-expanded="false"
                      class="dropdown-toggle btn btn-warning"
                      id="reject-account-request-0"
                      ngbdropdowntoggle=""
                      type="button"
                    >
                       Reject 
                    </button>
                    <div
                      class="dropdown-menu"
                      ngbdropdownmenu=""
                    >
                      <button
                        class="btn btn-light btn-sm dropdown-item"
                        id="reject-request-0"
                      >
                         Reject 
                      </button>
                      <button
                        class="btn btn-light btn-sm dropdown-item"
                        id="reject-request-with-reason-0"
                      >
                         Reject With Reason 
                      </button>
                    </div>
                  </span>
                </div>
              </td>
            </tr>
            <tr>
              <td>
                Instructor B
              </td>
              <td>
                instructorb@example.com
              </td>
              <td>
                APPROVED
              </td>
              <td>
                Institution and Country B
              </td>
              <td
                id="timestamp-box"
              >
                Created Time B
              </td>
              <td>
                <div
                  id="comment-box"
                >
                  Comment B
                </div>
              </td>
              <td
                class="align-middle"
              >
                <div
                  class="d-flex flex-row align-items-center justify-content-center gap-2"
                >
                  <div>
                    <a
                      id="edit-account-request-1"
                    >
                      <i
                        class="fa-solid fa-pen"
                      />
                    </a>
                  </div>
                  <div
                    class="ngb-tooltip-class"
                    placement="top"
                  >
                    <a
                      id="delete-account-request-1"
                    >
                      <i
                        class="fa-solid fa-trash"
                      />
                    </a>
                  </div>
                  <div>
                    <a
                      id="view-account-request-1"
                    >
                      <i
                        class="fa-solid fa-eye"
                      />
                    </a>
                  </div>
                  <button
                    class="btn btn-success"
                    disabled=""
                    id="approve-account-request-1"
                  >
                    Approve
                  </button>
                  <span
                    class="dropdown"
                    container="body"
                    ngbdropdown=""
                  >
                    <button
                      aria-expanded="false"
                      class="dropdown-toggle btn btn-warning"
                      disabled=""
                      id="reject-account-request-1"
                      ngbdropdowntoggle=""
                      type="button"
                    >
                       Reject 
                    </button>
                    <div
                      class="dropdown-menu"
                      ngbdropdownmenu=""
                    >
                      <button
                        class="btn btn-light btn-sm dropdown-item"
                        id="reject-request-1"
                      >
                         Reject 
                      </button>
                      <button
                        class="btn btn-light btn-sm dropdown-item"
                        id="reject-request-with-reason-1"
                      >
                         Reject With Reason 
                      </button>
                    </div>
                  </span>
                </div>
              </td>
            </tr>
            <tr>
              <td>
                Instructor C
              </td>
              <td>
                instructorc@example.com
              </td>
              <td>
                REJECTED
              </td>
              <td>
                Institution and Country C
              </td>
              <td
                id="timestamp-box"
              >
                Created Time C
              </td>
              <td>
                <div
                  id="comment-box"
                >
                  Comment C
                </div>
              </td>
              <td
                class="align-middle"
              >
                <div
                  class="d-flex flex-row align-items-center justify-content-center gap-2"
                >
                  <div>
                    <a
                      id="edit-account-request-2"
                    >
                      <i
                        class="fa-solid fa-pen"
                      />
                    </a>
                  </div>
                  <div
                    class="ngb-tooltip-class"
                    placement="top"
                  >
                    <a
                      id="delete-account-request-2"
                    >
                      <i
                        class="fa-solid fa-trash"
                      />
                    </a>
                  </div>
                  <div>
                    <a
                      id="view-account-request-2"
                    >
                      <i
                        class="fa-solid fa-eye"
                      />
                    </a>
                  </div>
                  <button
                    class="btn btn-success"
                    id="approve-account-request-2"
                  >
                    Approve
                  </button>
                  <span
                    class="dropdown"
                    container="body"
                    ngbdropdown=""
                  >
                    <button
                      aria-expanded="false"
                      class="dropdown-toggle btn btn-warning"
                      disabled=""
                      id="reject-account-request-2"
                      ngbdropdowntoggle=""
                      type="button"
                    >
                       Reject 
                    </button>
                    <div
                      class="dropdown-menu"
                      ngbdropdownmenu=""
                    >
                      <button
                        class="btn btn-light btn-sm dropdown-item"
                        id="reject-request-2"
                      >
                         Reject 
                      </button>
                      <button
                        class="btn btn-light btn-sm dropdown-item"
                        id="reject-request-with-reason-2"
                      >
                         Reject With Reason 
                      </button>
                    </div>
                  </span>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </tm-account-request-table>
</tm-admin-home-page>
`;
